prepreduce/prepreduce.py

```diff
@@ -0,0 +1,395 @@
+        cc_invocation=clean_cc_invocation,
+        output_dir=cli_args.output,
+        target=path_to_dot_i_file.with_suffix(".o"),
+        force_rm=cli_args.force_rm_existing_script,
+    )
+
+
+def write_test_script(
+    *, cc_invocation: list[str], output_dir: Path, target: Path, force_rm: bool
+) -> None:
+    """
+    Create a test.sh script at --output directory
+    """
+    script_path = Path(os.getcwd()) / Path("test.sh")
+    info(f"Generating {script_path}")
+    if script_path.exists():
+        double_check_removal_with_user(file=script_path, force_rm=force_rm)
+
+    output_target_idx = cc_invocation.index("-o") + 1
+    cc_invocation[output_target_idx] = str(target)
+
+    cc_flags = " ".join(cc_invocation[1:])
+    script_text = textwrap.dedent(
+        f"""    #! /usr/bin/sh
+    CC='clang'
+    EXTRA_FLAGS= \"$@\"
+    CC_invocation() {{
+        CC_FLAGS=\"{cc_flags}\"
+        CC_FLAGS=\"$EXTRA_FLAGS $CC_FLAGS\"
+        $CC $CC_FLAGS
+    }}
+    OUT=$(CC_invocation 2>&1)
+    echo $OUT | grep "<your test here>"
+    """
+    )
+
+    with open(script_path, "w") as fd:
+        fd.write(script_text)
+
+    success(f"Successfully wrote {script_path}")
+    try:
+        os.chmod(script_path, 0o755)
+        info(f"Added execute permissions to {script_path}")
+    except:
+        warn(
+            f"Couldn't change permissions of {script_path} to enable execution. "
+            "Do this manually."
+        )
+
+
+def double_check_removal_with_user(*, file: Path, force_rm: bool) -> None:
+    # TODO: add optional rename
+    while not force_rm:
+        user_inp = input(f"[🟡] target {file} already exists. Remove it? (y/n): ")
+        user_inp = user_inp.strip().lower()
+        if user_inp == "y":
+            break
+        elif user_inp == "n":
+            success("OK: quitting.")
+            exit(0)
+    info(f"Removing {file} which already exists")
+    file.unlink()
+
+
+def parse_cli_args() -> argparse.Namespace:
+    """
+    Parse out arguments from command line interface.
+    Use various validate_foo_from_cli functions to assist end-user
+    """
+    global DEBUG, VERSION
+
+    parser = argparse.ArgumentParser(
+        prog="prepreduce",
+        description=f"prepreduce v{VERSION}",
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
+    )
+
+    parser.add_argument(
+        "target",
+        help="The source file you wish to target. Example: lib/string.o",
+        type=validate_target_from_cli,
+    )
+
+    parser.add_argument(
+        "-p",
+        "--path-to-linux",
+        type=lambda arg: validate_path_from_cli(arg, "-p/--path-to-linux"),
+        default="./",
+        help="Specify the Linux source tree directory",
+    )
+
+    parser.add_argument(
+        "build_command",
+        help="The specific build command preamble used sans the target. Example: `make -j$(nproc) LLVM=1 V=1`",
+        nargs="*",
+        default=["make", "-j8", "LLVM=1", "V=1"],
```

Comment:
I wonder if it would be better just to have the user specify their variables and default to `f"-j{os.cpu_count()}"`.

Reply:
yeah 8 is too low of a default IMO.

At some point, we'll want to be able to specify `ARCH=` i.e. `ARCH=arm64`. Not necessary in V1, but something to keep in mind.